File: soonba/boj_1260.java
/*
 * BAEKJOON ONLINE JUDGE
 * https://www.acmicpc.net
 * Problem Number: 1260
 * Level: SILVER II
 * Algorithm: DFS, BFS
 */

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Collections;
import java.util.LinkedList;
import java.util.Queue;

/*

주어진 값에서 계속 사용되는 데이터는 static 으로 초기화한다.

nodeList
- 라인 정보를 가지고있는 배열로, 방향이 없으므로 한 라인에 2 노드에 모두 추가해준다.
- 같은 차수(뎁스?) 에서 우선순위를 부여하기 위해 정렬한다.

visitDFS
- dfs static 메서드 내에서 파라미터로 계속 넣지 않아도 됨
visitBFS
- 안해도 되지만 통일하고자 굳이 static..

dfs
재귀호출 방식으로 구현
시작 노드로 출발하여 방문 여부를 체크한 후 방문하지 않은 다음 노드에 대해 dfs 호출

bfs
큐를 이용한 while문으로 구현
시작 노드로 출발하여 큐가 빌 때까지 while문을 돈다. 큐에서 노드가 poll이 되면 해당 노드가 가진 다음 노드 중 방문하지 않은 노드를 넣는다.
 */
public class Main {
    static LinkedList<Integer>[] nodeList;
    static boolean[] visitBFS;
    static boolean[] visitDFS;
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] s = br.readLine().split(" ");
        int node = Integer.parseInt(s[0]);
        int lineInfo = Integer.parseInt(s[1]);
        int startNode = Integer.parseInt(s[2]);

        //노드 초기화
        nodeList = new LinkedList[node + 1];
        for (int i = 0; i < nodeList.length; i++) {
            nodeList[i] = new LinkedList<>();
        }
        for (int i = 0; i < lineInfo; i++) {
            String[] s1 = br.readLine().split(" ");
            int a = Integer.parseInt(s1[0]);
            int b = Integer.parseInt(s1[1]);
            //방향이 상관 없으므로 a, b 두 노드 모두 등록
            nodeList[a].add(b);
            nodeList[b].add(a);
        }

        // 작은 수부터 탐색하기 위해 정렬
        for (LinkedList<Integer> integers : nodeList) {
            Collections.sort(integers);
        }


        //dfs
        StringBuilder dfsResult = new StringBuilder();
        visitDFS = new boolean[node + 1];
        dfs(startNode,dfsResult);


        //bfs
        visitBFS = new boolean[node + 1];
        Queue<Integer> bfsQueue = new LinkedList<>();

        //첫 노드를 queue에 넣고 while문 시작
        visitBFS[startNode] = true;
        bfsQueue.add(startNode);

        StringBuilder bfsResult = new StringBuilder();
        while (!bfsQueue.isEmpty()) {

            Integer poll = bfsQueue.poll();
            bfsResult.append(poll).append(" ");

            //poll() 을 해주면서 해당 node의 모든 노드를 담는다. 이 때 이미 방문했다면 패스
            for (int i = 0; i < nodeList[poll].size(); i++) {
                Integer nextNode = nodeList[poll].get(i);
                if (!visitBFS[nextNode]) {
                    bfsQueue.add(nextNode);
                    visitBFS[nextNode] = true;
                }
            }
        }

        System.out.println(dfsResult);
        System.out.println(bfsResult);

    }

    static void dfs(int node, StringBuilder result) {
        //이미 방문한 노드라면 패스
        if(visitDFS[node]) {
            return;
        }

        //방문
        StringBuilder addedResult = result.append(node).append(" ");
        visitDFS[node] = true;
        for (int i = 0; i < nodeList[node].size(); i++) {
            //다음 노드 방문을 위한 재귀 호출
            Integer nextNode = nodeList[node].get(i);
            dfs(nextNode,addedResult);
        }
    }

}

File: perlpark/boj_1260.js
/*
 * BAEKJOON ONLINE JUDGE
 * https://www.acmicpc.net
 * Problem Number: 1260
 * Level: Silver II
 * Algorithm: Graph Theory / Graph Traversal / Breadth-first Search / Depth-first Search
 */

/* Pseudocode *
1. 입력의 조건과 정점의 정보를 받는다.
2. 그래프를 그린다. (g)
2-1. 정점의 개수 NxN 사이즈의 false로 채운 배열을 만든다.
2-2. 정점의 정보에 따라 간선이 놓인 정점 쌍에 true로 놓는다.
2-2-1. 배열의 인덱스는 0부터 시작하므로 정점의 정보에 -1을 해준다.
3. bfs
3-1. 큐에 시작 정점을 넣고 해당 정점을 확인했다는 표시로 visited로 설정한다. (vis 배열에 push 함)
3-1-1. 배열의 인덱스는 0부터 시작하므로 시작 정점의 번호에 -1을 해준다.
3-2. 큐가 빌 때까지 반복하는 반복문을 선언한다.
3-3. 반복문에서 큐의 첫 번째 정점을 꺼낸다. (cur) 해당 정점을 결과에 포함 (result에 추가)
3-3-1. 정점 번호에 -1을 했으므로 결과에 저장할 떈 다시 +1을 해준다.
3-4. 해당 정점의 간선 정보를 보기 위해 그래프 g[cur]을 순회하는 반복문을 선언한다.
3-5. 다른 정점(i)이 이미 확인한 적 있다면 패스
3-6. 다른 정점(i)이 현재 정점(cur)과 연결되어있지 않아도 패스
3-7. 다른 정점(i)이 현재 정점(cur)과 연결되어 있고 확인한 적 없다면 큐에 넣고, 확인했다는 표시로 visited로 설정한다. (vis 배열에 push 함)
3-8. 최종적으로 큐가 다 비면 result를 출력하도록 한다.
4. dfs
4-1. dfs는 함수를 재귀하여, 방문 표시와 result에 담을 때 연결되어 있는 정점들로 순서를 우선한다는 차이만 있따....
예를 들면 bfs는 cur -> cur과 연결된 다른 정점들을 우선 vis,
        dfs는 cur -> cur과 연결된 다른 정점 1개(*) vis -> 그 다른 정점과 연결된 다른 정점 1개 vis -> 연결된 정점이 없으면 이전 정점으로 돌아감
        이런 차이가 있다...

*) 방문할 수 있는 정점이 여러 개인 경우에는 정점 번호가 작은 것을 먼저 방문한다는 조건인 경우, 연결된 정점 중 번호가 낮은 것이 선택된다.

5. dfs -> bfs 순으로 호출한다.
*/

const [condition, ...m] = `${require('fs').readFileSync('/dev/stdin')}`.trim().split(/\n/);

const [N, M, V] = condition.split(/\s/);
const newArray = (f) => new Array(+N).fill(f);

const g = [];
const start = +V - 1;

// 그래프 생성
for (let i = 0; i < +N; i++) {
  g[i] = newArray(false);
}
for (let i = 0; i < +M; i++) {
  const [start, end] = m[i].split(' ').map((v) => v - 1);
  g[start][end] = true;
  g[end][start] = true;
}

// BFS
function bfs() {
  const q = [];
  const vis = newArray(false);
  let result = '';

  q.push(start);
  vis[start] = true;

  while (q.length > 0) {
    const cur = q.shift();
    result += cur + 1 + ' ';

    for (let i = 0; i < +N; i++) {
      if (vis[i]) continue;
      if (!g[cur][i]) continue;

      q.push(i);
      vis[i] = true;
    }
  }
  console.log(result);
}

// DFS
function dfs() {
  const vis = newArray(false);
  let result = '';

  function search(cur) {
    vis[cur] = true;
    result += cur + 1 + ' ';

    for (let i = 0; i < +N; i++) {
      if (vis[i]) continue;
      if (!g[cur][i]) continue;

      search(i);
    }
  }
  search(start);
  console.log(result);
}

dfs();
bfs();
